examples/features/gracefulstop/server/main.go

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Binary server demonstrates how to gracefully stop a gRPC server.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"google.golang.org/grpc"
+	pb "google.golang.org/grpc/examples/features/proto/echo"
+)
+
+var (
+	port           = flag.Int("port", 50052, "port number")
+	streamMessages int32
+	mu             sync.Mutex
+)
+
+type server struct {
+	pb.UnimplementedEchoServer
+}
+
+// ServerStreamingEcho implements the EchoService.ServerStreamingEcho method.
+// It receives an EchoRequest and sends back a stream of EchoResponses.
+// The stream will contain up to 5 messages, each sent after a 1-second delay.
+// If the client cancels the request or if more than 5 messages are sent,
+// the stream will be closed with an error.
+func (s *server) ServerStreamingEcho(_ *pb.EchoRequest, stream pb.Echo_ServerStreamingEchoServer) error {
+	ctx := stream.Context()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			atomic.AddInt32(&streamMessages, 1)
+
+			mu.Lock()
+			if streamMessages > 5 {
+				return fmt.Errorf("request failed")
+			}
```

Comment:
Why return an error?  Exiting successfully feels more graceful.

Also you should _always_ return status errors from your method handlers.  This will get converted to UNKNOWN by gRPC, which "is fine", but we should never show an example that does anything but send statuses back.

Reply:
Not applicable now. Explained the flow above